target/linux/ath79/dts/ar7242_meraki_mr12.dts

```diff
@@ -0,0 +1,180 @@
+#include "ar7242.dtsi"
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+/ {
+	compatible = "meraki,mr12", "qca,ar7242";
+	model = "Meraki MR12";
+
+	aliases {
+		led-boot = &led_power_orange;
+		led-failsafe = &led_power_orange;
+		led-running = &led_power_green;
+		led-upgrade = &led_power_orange;
+	};
+
+	extosc: ref {
+		compatible = "fixed-clock";
+		#clock-cells = <0>;
+		clock-output-names = "ref";
+		clock-frequency = <40000000>;
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		wifi1 {
+			label = "green:wifi1";
+			gpios = <&gpio 11 GPIO_ACTIVE_LOW>;
+		};
+
+		wifi2 {
+			label = "green:wifi2";
+			gpios = <&gpio 12 GPIO_ACTIVE_LOW>;
+		};
+
+		wifi3 {
+			label = "green:wifi3";
+			gpios = <&gpio 13 GPIO_ACTIVE_LOW>;
+		};
+
+		wifi4 {
+			label = "green:wifi4";
+			gpios = <&gpio 14 GPIO_ACTIVE_LOW>;
+		};
+
+		wan {
+			label = "green:wan";
+			gpios = <&gpio 15 GPIO_ACTIVE_LOW>;
+		};
+
+		led_power_orange: power_orange {
+			label = "orange:power";
+			gpios = <&gpio 16 GPIO_ACTIVE_LOW>;
+			panic-indicator;
+		};
+
+		led_power_green: power_green {
+			label = "green:power";
+			gpios = <&gpio 17 GPIO_ACTIVE_LOW>;
+		};
+	};
+
+	keys {
+		compatible = "gpio-keys";
+
+		reset {
+			label = "reset";
+			linux,code = <KEY_RESTART>;
+			gpios = <&gpio 8 GPIO_ACTIVE_LOW>;
+		};
+	};
+};
+
+&pcie {
+	status = "okay";
+
+	ath9k: wifi@0,0 {
+		reg = <0x0000 0 0 0 0>;
+		qca,no-eeprom;
+		mtd-mac-address = <&config 0x66>;
+		mtd-mac-address-increment = <1>;
+		#gpio-cells = <2>;
+		gpio-controller;
+	};
+};
+
+&pll {
+	clocks = <&extosc>;
+};
+
+&uart {
+	status = "okay";
+};
+
+&mdio0 {
+	status = "okay";
+
+	phy4: ethernet-phy@4 {
+		reg = <4>;
+	};
+};
+
+&eth0 {
+	status = "okay";
+	mtd-mac-address = <&config 0x66>;
+
+        pll-data = <0x16000000 0x00000101 0x00001313>;
+        pll-reg = <0x4 0x2c 17>;
+        pll-handle = <&pll>;
+
+	phy-mode = "rgmii-id";
+	phy-handle = <&phy4>;
+};
+
+&mdio1 {
+	status = "okay";
+
+	phy0: ethernet-phy@0 {
+		reg = <0>;
+	};
+};
```

Comment:
I'm assuming I will still need the `&eth1 { ... mtd-mac-address ... }`, so here's the patch I've tried:

```diff
diff --git a/target/linux/ath79/dts/ar7242_meraki_mr12.dts b/target/linux/ath79/dts/ar7242_meraki_mr12.dts
index 3ecbd620d9..969666ea9a 100644
--- a/target/linux/ath79/dts/ar7242_meraki_mr12.dts
+++ b/target/linux/ath79/dts/ar7242_meraki_mr12.dts
@@ -113,23 +113,9 @@
        phy-handle = <&phy4>;
 };
 
-&mdio1 {
-       status = "okay";
-
-       phy0: ethernet-phy@0 {
-               reg = <0>;
-       };
-};
-
 &eth1 {
        status = "okay";
        mtd-mac-address = <&config 0x66>;
-
-       phy-handle = <&phy0>;
-        fixed-link {
-               speed = <100>;
-               full-duplex;
-       };
 };
 
 &spi {
```

It does bring up a port, but it's sadly still set to 1000Mbit/s, probably due to this in `ar7242.dtsi`: 

```devicetree
&eth1 {
	compatible = "qca,ar7242-eth", "syscon";

	resets = <&rst 13>;
	reset-names = "mac";

	phy-mode = "gmii";

	fixed-link {
		speed = <1000>;
		full-duplex;
	};
};
```

I switched the connection to my laptop (T480s) and while I could negotiate a 100M connection, this was not reflected on the MR12's side, and I received nothing back from an DHCPv4, despite eth1 being bridged into my home LAN via eth0.

More information:

```sh
root@OpenWrt:/# dmesg | grep 9000 -C 2
[    1.116476] random: fast init done
[    1.308953] libphy: ag71xx_mdio: probed
[    1.341770] ag71xx 19000000.eth: connected to PHY at mdio.0:04 [uid=004dd04e, driver=Generic PHY]
[    1.351688] eth0: Atheros AG71xx at 0xb9000000, irq 4, mode: rgmii-id
[    1.701427] ag71xx 1a000000.eth: connected to PHY at fixed-0:00 [uid=00000000, driver=Generic PHY]
[    1.711479] eth1: Atheros AG71xx at 0xba000000, irq 5, mode: gmii
root@OpenWrt:/# ethtool eth1
Settings for eth1:
        Supported ports: [ TP MII ]
        Supported link modes:   1000baseT/Full 
        Supported pause frame use: Symmetric Receive-only
        Supports auto-negotiation: Yes
        Supported FEC modes: Not reported
        Advertised link modes:  Not reported
        Advertised pause frame use: No
        Advertised auto-negotiation: Yes
        Advertised FEC modes: Not reported
        Link partner advertised link modes:  1000baseT/Full 
        Link partner advertised pause frame use: No
        Link partner advertised auto-negotiation: No
        Link partner advertised FEC modes: Not reported
        Speed: 100Mb/s
        Duplex: Full
        Port: MII
        PHYAD: 0
        Transceiver: external
        Auto-negotiation: on
        Current message level: 0x000000ff (255)
                               drv probe link timer ifdown ifup rx_err tx_err
        Link detected: yes
root@OpenWrt:/# ethtool eth0
Settings for eth0:
        Supported ports: [ TP MII ]
        Supported link modes:   10baseT/Half 10baseT/Full 
                                100baseT/Half 100baseT/Full 
                                1000baseT/Full 
        Supported pause frame use: Symmetric Receive-only
        Supports auto-negotiation: Yes
        Supported FEC modes: Not reported
        Advertised link modes:  1000baseT/Full 
        Advertised pause frame use: No
        Advertised auto-negotiation: Yes
        Advertised FEC modes: Not reported
        Link partner advertised link modes:  10baseT/Half 10baseT/Full 
                                             100baseT/Half 100baseT/Full 
                                             1000baseT/Full 
        Link partner advertised pause frame use: Symmetric Receive-only
        Link partner advertised auto-negotiation: Yes
        Link partner advertised FEC modes: Not reported
        Speed: 1000Mb/s
        Duplex: Full
        Port: MII
        PHYAD: 4
        Transceiver: external
        Auto-negotiation: on
        Current message level: 0x000000ff (255)
                               drv probe link timer ifdown ifup rx_err tx_err
        Link detected: yes
```

I wonder if it'd be possible to "overwrite" the eth1 from `ar7242.dtsi` to allow autonegotiation to do its thing.

Reply:
you can't change the speed from the MR12 for eth1 because that is the internal switch that has is a fixed link, like how it is set in the dtsi

you don't have to worry about the different speeds for eth1, just make sure it pings, etc.

Reply:
> you can't change the speed from the MR12 for eth1 because that is the internal switch that has is a fixed link, like how it is set in the dtsi. ethtool can only control the link between the switch and the SOC.
> 
> you don't have to worry about the different speeds for eth1, just make sure it pings, etc.

Looking at page 24 of the spec sheet I can see the reasoning for eth1 using gmii + fixed-link, makes sense. I just can't ping anything over it. either with 100 or ar7242's 1000. I thought I was able to do so before but I'm not so sure now.

> and if it doesn't, you use swconfig to see which port number links up and add the switch configuration to 02_network

This one is probably it. I'll check how it's done with others ...

Reply:
disconnect ethernet

`swconfig dev switch0 show`

connect ethernet

`swconfig dev switch0 show`

you would see 1 link up before, 2 links up after

depending on the number something like this

```
		ucidef_set_interface_lan "eth0"
		ucidef_add_switch "switch0" \
			"0@eth1" "X:lan:1"

```

where you replace X with the port number

Reply:
It actually doesn't detect the switch if &mdio1 isn't `okay`ed, but once I do that and add a `ucidef_add_switch` entry in `02_network`, everything works. Woo!

Tested the switch with 100 and 10 speeds on the external port 1 of the switch; both work.

Reply:
@adschm do you think this is a bug? I have never been sure what to do with mdio1 node when nothing needs to be changed...

Reply:
This does not make much sense to me. I'm not a switch setup expert, but from looking at mr12/mr16 in ar71xx and from looking at other ar7242 devices, I don't see a reason for touching mdio1/eth1 for this device.
I'm also not sure whether this is a mistake, but the 02_network setup also contains _two_ eths, although this is supposed to only have one port?
Me, I'd rather expect something very close to mr16, just with phy4 instead of phy0 ...

Reply:
Apparently mr12 in ar71xx never had the second port at all....

Reply:
> Apparently mr12 in ar71xx never had the second port at all....
> 
> according to FCC photos its a 2 port device
> https://fccid.io/UDX-60013010/Internal-Photos/Internal-Photos-1395141.pdf

@riptidewave93 claims he had issues getting the switch to work due to some driver issue. I'm not sure what that was about, but it's definitely a two-port device.

> the possible bug I am talking about is that it should not be necessary to have mdio1 node in dts if mdio1 does not need to be changed, because it is not disabled in dtsi

I can provide the fdt that gets built with and without the mdio1 mentioned in the .dts if you want. All I know is that `swconfig list` shows no switches when `mdio1` is not mentioned in `ar7242_meraki_mr12.dts`.

Reply:
@adschm do you think mdio1 in the dtsi should have `status = "okay";`?

Reply:
Built just now without okay'ing `mdio1`:

```sh
root@OpenWrt:/# swconfig list
root@OpenWrt:/# dmesg | head -1
[    0.000000] Linux version 5.4.80 (labby@lobon) (gcc version 8.4.0 (OpenWrt GCC 8.4.0 r14947-0f7a3288e1)) #0 Sun Nov 29 00:29:40 2020
root@OpenWrt:/# dmesg | grep ag71
[    1.304987] libphy: ag71xx_mdio: probed
[    1.338026] ag71xx 19000000.eth: connected to PHY at mdio.0:04 [uid=004dd04e, driver=Generic PHY]
[    1.701369] ag71xx 1a000000.eth: connected to PHY at fixed-0:00 [uid=00000000, driver=Generic PHY]
root@OpenWrt:/# 
```

Building again with mdio1 okay'ed ...

Reply:
Built once more with okay'ed `mdio1`:

```sh
root@OpenWrt:/# swconfig list
Found: switch0 - mdio.1:1f
root@OpenWrt:/# dmesg | head -1
[    0.000000] Linux version 5.4.80 (labby@lobon) (gcc version 8.4.0 (OpenWrt GCC 8.4.0 r14947-0f7a3288e1)) #0 Sun Nov 29 00:29:40 2020
root@OpenWrt:/# dmesg | grep ag71
[    1.304946] libphy: ag71xx_mdio: probed
[    1.337973] ag71xx 19000000.eth: connected to PHY at mdio.0:04 [uid=004dd04e, driver=Generic PHY]
[    1.988956] libphy: ag71xx_mdio: probed
[    2.229776] ag71xx 1a000000.eth: connected to PHY at fixed-0:00 [uid=00000000, driver=Generic PHY]
root@OpenWrt:/# 
```

It looks like okay'ing `mdio1` helps somehow. Could it be a `ucidef` bug?

Reply:
> @adschm do you think mdio1 in the dtsi should have status = "okay";?

Not in the DTSI, since other devices with one port do not need it.